### DISCUSSION TITLE
Sorting out USB HID and USB MIDI support for RP2040

### DISCUSSION THREAD

**Discussion by ayjaym** [micropython/micropython#11782] [RP2040 / Pico] [answered]
Hi folks
I know this has been discussed elsewhere and I appreciate work has been done in branches but its unclear if and when this will actually make it into an actual release.
There are a *lot* of useful projects that critically depend on having this functionality and having to resort to CircuitPython isn't really an option because then you lose threading support (and CP timing issues will then also be a problem).

I can't quite understand why the Raspberry Pi Foundation aren't helping out here, to be honest. With continued supply issues with other Pi Models, the Pico is a very attractive controller and has potential in a lot of scenarios. As it stands I have a major open source hardware project completely stalled because of this issue and I could certainly attempt to integrate the work done by others but discussions seem to indicate that due to ongoing Micropython refactoring would make that pointless, and I would also have to stand up a fork somewhere for my project users, which is awkward. Is there some way to expedite this work to get it into the main trunk in the not too distant future?

**Comment by jimmo** [comment]
Work is in progress here: https://github.com/micropython/micropython/pull/9497
Please feel free to get involved there.

MicroPython is run by a small team, we have a lot of different priorities. USB device support (and soon, host support too) is a major feature that has a lot of decisions to be made around APIs etc that we will be stuck with for a long time, so it takes time.

**Reply by ayjaym** [reply]
Thank you - I totally understand and my comments aren't made as criticisms - I know as a senior architect myself that some enhancements have deep implications architecturally and have to be done right or they can become a major problem in the future. Indeed, Python itself is proof of this with the 2.7->3.x breaking changes that tidied up aspects of the language that in hindsight were not optimal choices.

In thinking about this my obvious solution is to remember that 'the perfect is the enemy of the good'. As a consequence what I will do is replace the Pico with a Pico W and use rtpMIDI (which I have previously used) to send and receive MIDI wirelessly. In my experience latency is fairly acceptable and jitter is USUALLY ok - some WiFi adaptors can behave badly but this was more of a problem a few years ago with crap drivers than it is now.

This will let me at least publish this project because I'm quite excited about the possibilities. It allows the user to build a MIDI controller with more than 128 buttons, each RGB led backed, with micro SD reader, legacy MIDI out and an OLED display with rotary encoder, then use MicroPython to program it to do almost anything - a sequencer, or perhaps just 128 buttons to change patches, or DMX lighting control etc. etc. and the total build cost is well under £100. 

Anyway - I definitely am interested also in looking at ongoing work with wired USB host support - I am an experienced C++ developer as well as fairly familiar now with Python - however I was put off by the comments that header file changes etc. would make current work likely to be moot, so I was tempted to hold off until the broad architectural decisions have been made by the team and the underlying infrastructure looks stable. 

Ideally having both HID and MIDI support and the ability to concurrently still debug over USB serial is clearly very desirable, with the ability to leverage the dual core RP2040 so that USB-related stuff is handled on a dedicated core. 

I then have a bunch of stuff I'd like to move to the Pico - a MIDI footpedal that used the Pi Zero being an obvious choice - this lets the user control e.g a PDF reader with a musical score to move forward and backward through the score. You can purchase devices to do this but they are relatively expensive and not really configurable. The Pi Zero solution suffers - like all mainstream Pi devices - with lengthy boot times, whereas the Pico is obviously up and running almost instantly.

Also I have some ideas for Flight Simulator devices etc - there's a cool 8 rotary encoder module that you can buy for £15 that has RGB leds for each encoder and reads over I2C that would make a great controller for HID/MIDI using the Pico - I have purchased the module but again I'm stuck on the USB host issue for now.

Anyway - I do know you have limited resources - I guess if people can post as and when major decisions have been baked into the roadmap that would be very useful. As I said, I kinda wonder why Mr Upton and his team don't provide some support - you folks are doing great work but it ultimately benefits the Pi Foundation arguably more than just about any other vendor given their unit sales.

**Comment by jonnor** [comment]
@ayjaym USB device support has now been merged (see MR that jimmo linked). It should be released in 1.23

**Reply by sherif-audibene** [reply]
@jonnor does it support midi as well, Thanks, I looked in the example and couldn't find something

**Reply by scruss** [reply, marked as answer]
> does it support midi as well, Thanks, I looked in the example and couldn't find something

Yes, there's a MIDI library and example here: [micropython/usb](https://github.com/micropython/micropython-lib/tree/master/micropython/usb)

**Reply by coreyker** [reply]
What is the best way to install this package? I tried mip and that doesn't work (I'm guessing it's not indexed)

**Reply by projectgus** [reply]
@coreyker mip should be able to install these. What command are you running and what output do you get?

**Comment by sosi-deadeye** [comment]
Documentation is not everyone's strong point. However, if examples have been documented, you should definitely read them.

https://github.com/micropython/micropython-lib/blob/master/micropython/usb/examples/device/midi_example.py

```
# This example demonstrates creating a custom MIDI device.
#
# To run this example:
#
# 1. Make sure `usb-device-midi` is installed via: mpremote mip install usb-device-midi
#
# 2. Run the example via: mpremote run midi_example.py
#
# 3. mpremote will exit with an error after the previous step, because when the
#    example runs the existing USB device disconnects and then re-enumerates with
#    the MIDI interface present. At this point, the example is running.
#
# 4. To see output from the example, re-connect: mpremote connect PORTNAME
```

**Reply by coreyker** [reply]
I had messed up the mpremote command slightly but copy-pasting the one here worked. Thanks for the help and sorry for the noise!
